// mui-file-input/dist/mui-file-input.es.js
import { jsxs as h, jsx as s } from "react/jsx-runtime";
import p from "react";
import { styled as N } from "@mui/material/styles";
import R from "@mui/material/IconButton";
import b from "@mui/material/InputAdornment";
import U from "@mui/material/TextField";
import G from "@mui/material/Typography";
const Z = [
  "B",
  "kB",
  "MB",
  "GB",
  "TB",
  "PB",
  "EB",
  "ZB",
  "YB"
], _ = [
  "B",
  "KiB",
  "MiB",
  "GiB",
  "TiB",
  "PiB",
  "EiB",
  "ZiB",
  "YiB"
], j = [
  "b",
  "kbit",
  "Mbit",
  "Gbit",
  "Tbit",
  "Pbit",
  "Ebit",
  "Zbit",
  "Ybit"
], O = [
  "b",
  "kibit",
  "Mibit",
  "Gibit",
  "Tibit",
  "Pibit",
  "Eibit",
  "Zibit",
  "Yibit"
], F = (t, i, e) => {
  let n = t;
  return typeof i == "string" || Array.isArray(i) ? n = t.toLocaleString(i, e) : (i === !0 || e !== void 0) && (n = t.toLocaleString(void 0, e)), n;
};
function A(t, i) {
  if (!Number.isFinite(t))
    throw new TypeError(`Expected a finite number, got ${typeof t}: ${t}`);
  i = {
    bits: !1,
    binary: !1,
    space: !0,
    ...i
  };
  const e = i.bits ? i.binary ? O : j : i.binary ? _ : Z, n = i.space ? " " : "";
  if (i.signed && t === 0)
    return ` 0${n}${e[0]}`;
  const r = t < 0, c = r ? "-" : i.signed ? "+" : "";
  r && (t = -t);
  let l;
  if (i.minimumFractionDigits !== void 0 && (l = { minimumFractionDigits: i.minimumFractionDigits }), i.maximumFractionDigits !== void 0 && (l = { maximumFractionDigits: i.maximumFractionDigits, ...l }), t < 1) {
    const f = F(t, i.locale, l);
    return c + f + n + e[0];
  }
  const u = Math.min(Math.floor(i.binary ? Math.log(t) / Math.log(1024) : Math.log10(t) / 3), e.length - 1);
  t /= (i.binary ? 1024 : 1e3) ** u, l || (t = t.toPrecision(3));
  const g = F(Number(t), i.locale, l), d = e[u];
  return c + g + n + d;
}
const W = N("label")`
  position: relative;
  flex-grow: 1;

  input {
    opacity: 0 !important;
  }

  & > span {
    position: absolute;
    left: 0;
    right: 0;
    top: 0;
    bottom: 0;
    z-index: 2;
    display: flex;
    align-items: center;
  }

  span.MuiFileInput-placeholder {
    color: gray;
  }
`, H = N("div")`
  display: flex;
  width: 100%;

  & > span {
    display: block;
  }

  & > span:first-of-type {
    white-space: nowrap;
    text-overflow: ellipsis;
    overflow: hidden;
  }

  & > span:last-of-type {
    flex-shrink: 0;
    display: block;
  }
`, M = {
  Label: W,
  Filename: H
}, K = ({ text: t, isPlaceholder: i, placeholder: e, ...n }, r) => /* @__PURE__ */ h(M.Label, { children: [
  /* @__PURE__ */ s("input", { ...n, ref: r }),
  t ? /* @__PURE__ */ s(
    "span",
    {
      "aria-placeholder": e,
      className: i ? "MuiFileInput-placeholder" : "",
      children: typeof t == "string" ? t : /* @__PURE__ */ h(M.Filename, { children: [
        /* @__PURE__ */ s("span", { children: t.filename }),
        /* @__PURE__ */ h("span", { children: [
          ".",
          t.extension
        ] })
      ] })
    }
  ) : null
] }), V = p.forwardRef(K);
function q(t) {
  return t.length > 0;
}
function J(t) {
  return t.reduce((i, e) => i + e.size, 0);
}
function B(t) {
  return typeof window < "u" && t instanceof File;
}
function Q(t) {
  return Array.from(t);
}
function X(t) {
  const e = (B(t) ? t.name : t[0]?.name || "").split("."), n = e.pop();
  return {
    filename: e.join("."),
    extension: n
  };
}
const tt = typeof window < "u" ? p.useLayoutEffect : p.useEffect, at = p.forwardRef(
  (t, i) => {
    const {
      value: e,
      onChange: n,
      disabled: r,
      getInputText: c,
      getSizeText: l,
      placeholder: u,
      hideSizeText: g,
      inputProps: d,
      InputProps: f,
      multiple: I,
      className: P,
      clearIconButtonProps: S = {},
      ...E
    } = t, { className: w = "", ...v } = S, m = p.useRef(null), { startAdornment: C, ...z } = f || {}, L = I || d?.multiple || f?.inputProps?.multiple || !1, T = () => {
      m.current && (m.current.value = "");
    }, D = (o) => {
      const x = o.target.files, y = x ? Q(x) : [];
      I ? (n?.(y), y.length === 0 && T()) : (n?.(y[0] || null), y[0] || T());
    }, $ = (o) => {
      o.preventDefault(), !r && n?.(I ? [] : null);
    }, a = Array.isArray(e) ? q(e) : B(e);
    tt(() => {
      const o = m.current;
      o && !a && (o.value = "");
    }, [a]);
    const k = () => e === null || Array.isArray(e) && e.length === 0 ? u || "" : typeof c == "function" && e !== void 0 ? c(e) : e && a ? Array.isArray(e) && e.length > 1 ? `${e.length} files` : X(e) : "", Y = () => {
      if (typeof l == "function" && e !== void 0)
        return l(e);
      if (a) {
        if (Array.isArray(e)) {
          const o = J(e);
          return A(o);
        }
        if (B(e))
          return A(e.size);
      }
      return "";
    };
    return /* @__PURE__ */ s(
      U,
      {
        ref: i,
        type: "file",
        disabled: r,
        onChange: D,
        className: `MuiFileInput-TextField ${P || ""}`,
        InputProps: {
          startAdornment: /* @__PURE__ */ s(b, { position: "start", children: C }),
          endAdornment: /* @__PURE__ */ h(
            b,
            {
              position: "end",
              style: { visibility: a ? "visible" : "hidden" },
              children: [
                g ? null : /* @__PURE__ */ s(
                  G,
                  {
                    variant: "caption",
                    mr: "2px",
                    lineHeight: 1,
                    className: "MuiFileInput-Typography-size-text",
                    children: Y()
                  }
                ),
                /* @__PURE__ */ s(
                  R,
                  {
                    "aria-label": "Clear",
                    title: "Clear",
                    size: "small",
                    disabled: r,
                    className: `${w} MuiFileInput-ClearIconButton`,
                    onClick: $,
                    ...v
                  }
                )
              ]
            }
          ),
          ...z,
          inputProps: {
            text: k(),
            multiple: L,
            ref: m,
            isPlaceholder: !a,
            placeholder: u,
            ...d,
            ...f?.inputProps
          },
          // @ts-expect-error
          inputComponent: V
        },
        ...E
      }
    );
  }
);
export {
  at as MuiFileInput
};
